### PR TITLE
WIP dictdiffer: diff methods and merging

### PR DESCRIPTION
**Pull request by mvesper**
* This commit is not in a working state (at least not tested), it is just
  there to support a discussion.

* Adds the option for different diff algorithms.
  Because certain difference algorithms would require the patches to be
  in order, the 'if differ:...' part had to be reworked, which led to
  patches generally having only one value instead of a list of them.

* Adds the option to expand the patch extraction. Also adds an variable
  to decide on the depth of expansion (path_limit).

* Adds merging functionality, therefor reworking the structure of the
  package.

* The current changes are not fully tested.

# UPDATE:
* There are now tests (there are still known issues)
* Restructured the dictdiffer module

# To get an idea of the possible workflow:

## The default way
```
from dictdiffer import diff, merge

lca = {'foo': 'bar',
       'insert_list': [3],
       'change_list': [0,1,2,3,4,5,6]}

new1 = {'foo': 'foo-bar',
        'insert_list': [1,2,3,4,5],
        'change_list': [0,-1,-2,-3,-4,-5,6]}

new2 = {'foo': 'bar-foo',
        'insert_list': [1,2,3],
        'change_list': [0,-1,2,3,4,-5,6]}

diff1 = list(diff(lca, new1))
diff2 = list(diff(lca, new2))

success, patches = merge(diff1, diff2)
```

## The configurable way
```
from dictdiffer import diff, merge
from dictdiffer.utils import WildcardDict
from dictdiffer.extractors import get_default_extractors, SequenceExtractor
from dictdiffer.wrap.wrappers import SequenceWrapper

lca = {'foo': 'bar',
       'insert_list': [3],
       'change_list': [0,1,2,3,4,5,6]}

new1 = {'foo': 'foo-bar',
        'insert_list': [1,2,3,4,5],
        'change_list': [0,-1,-2,-3,-4,-5,6]}

new2 = {'foo': 'bar-foo',
        'insert_list': [1,2,3],
        'change_list': [0,-1,2,3,4,-5,6]}


e = get_default_extractors()
e[('insert_list',)] = SequenceExtractor
e[('change_list',)] = SequenceExtractor

diff1 = list(diff(lca, new1, expand=True, extractors=e)) 
diff2 = list(diff(lca, new2, expand=True, extractors=e)) 

# Attention, those wrappers might end up storing some internal data, using
# the same dicitonary for both wrappers like w1 = w1 = WildcardDict(...)
# is probably going to be a bad idea
wrapper1 = WildcardDict({('insert_list', '*'): SequenceWrapper(),
                         ('change_list', '*'): SequenceWrapper()})

wrapper2 = WildcardDict({('insert_list', '*'): SequenceWrapper(),
                         ('change_list', '*'): SequenceWrapper()})

def take_left(patch1, patch2, patches1, patches2, ai):
    patch1['take'] = range(len(patch1['patches']))
    patch2['take'] = []

    return True

actions = WildcardDict({('foo',): take_left,
                        ('change_list', '*'): take_left})

success, patches = merge(diff1, diff2, actions, wrapper1, wrapper2)
```

# UPDATE:
This pull request will be closed, since a more pragmatic approach was implemented in #53.

### CODE REVIEW

**Comment by jirikuncar** [comment]
@mvesper do you have any updates?

**Comment by mvesper** [comment]
@jirikuncar I worked on tests recently and I am making some more adjustments there. I will try update the pull request soon.

**Comment by mvesper** [comment]
@jirikuncar 
>should be enough to have only dictdiffer/wrappers.py

Are you sure about this? The wrapping seems to be a distinct part of the merging process (to be more precise: the preparation of it). It feels like if I move this I should also move other files like unify.

**Comment by jirikuncar** [comment]
Why `dictdiffer.wrappers` is not enough? I'm not sure if I understand your concern here. See also my other https://github.com/inveniosoftware/dictdiffer/pull/46#commitcomment-8947086.

File: dictdiffer/__init__.py [hunk 1]
```diff
@@ -9,23 +9,33 @@
...
 """Dictdiffer is a helper module to diff and patch dictionaries."""
 
-import sys
 import copy
 
-if sys.version_info[0] == 3:  # pragma: no cover (Python 2/3 specific code)
-    string_types = str,
-else:  # pragma: no cover (Python 2/3 specific code)
-    string_types = basestring,
+from utils import (WildcardDict,
```
Comment (jirikuncar) [inline dictdiffer/__init__.py]: I would be nice to use relative imports everywhere (`.utils`, `.wrappers`, etc.) and sort imports according to PEP8.

File: dictdiffer/__init__.py [hunk 2]
```diff
@@ -43,82 +53,89 @@ def diff(first, second, node=None, ignore=None):
...
+                        return extractor
 
-        intersection = {}
-        addition = second - first
-        if len(addition):
-            yield ADD, dotted_node, [(0, addition)]
-        deletion = first - second
-        if len(deletion):
-            yield REMOVE, dotted_node, [(0, deletion)]
+        return None
```
Comment (jirikuncar) [inline dictdiffer/__init__.py]: This is kind of explicit. Please consider if it wouldn't be better to remove it.

**Comment by kaplun** [comment]
@mvesper, @jirikuncar what is missing of this PR to make it through?

**Comment by jirikuncar** [comment]
@kaplun @mvesper 

Just some notes with things to be checked:
- [ ] add documentation about `.. versionchanged:: 1.0.0` in `diff` function describing incompatible changes and new arguments;
- [ ] rebase on top *master* and keep all tests from `0.x` version untouched when possible;
- [ ] reformat commit message;
- [ ] ask more people for reviews (cc @lnielsen)

**Comment by mvesper** [comment]
@kaplun, @jirikuncar 

So, it took quite some time, but since my master thesis is finished now I would like to talk about this pull request in a more pragmatic way:

For the sake of my thesis I introduced some generality and features which, after some experimentation, turned out to be not needed to solve the problem that originated this work.

It became apparent that it is easier to pass the references or author lists as a whole to a resolving function, which means that we don't need to introduce another difference algorithm which is based on some kind of sequence extraction and that the current difference extraction process for lists is suitable enough or will never be used in our case.

This means: We could introduce the possibility for custom extraction algorithms just for the sake of it and risk that it breaks some users code without providing us any value. Even though this pull request does not break the usage of the dictdiffers API, it seems quite possible that it might break some code because it introduces an order to the patches, which changes their layout a bit (Each deletion or insertion patch only contains one element in the list of changes (third element of the tuple)).

As I told @kaplun, I am currently figuring out how much time and effort it would take to make the merging process work with the current style patches. Up until now it is mostly about getting rid of the most tedious code changes.

So, comments? Do we want to give users the option to use their own extraction algorithms or not?

**Comment by mvesper** [comment]
So, I don't know if I should create a new, competing pull request or not, but:

I created a more 'pragmatic' approach for the merging, which you can find here 363c6938dde870f93f2edc16373850da01ea9bc8

(I tested the implementation only roughly)

Tell me what you think is a more reasonable approach.

**Comment by jirikuncar** [comment]
I like the implementation with `expand=True`.

**Comment by mvesper** [comment]
This pull request will be closed, since a more pragmatic approach was implemented in #53.

**Pull request closed without merging**